akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala

```diff
@@ -753,6 +766,86 @@ class Http2ClientSpec extends AkkaSpecWithMaterializer("""
+
+      "until in-flight responses are received" inAssertAllStagesStopped new TestSetup with NetProbes {
+        // Given an in-flight request...
+        user.emitRequest(HttpRequest(uri = "https://www.example.com/"))
+        network.expectDecodedResponseHEADERSPairs(1)
+        // ... and a completed user handler
+        user.requestOut.sendComplete()
+
+        network.sendFrame(
+          HeadersFrame(streamId = 1, endStream = true, endHeaders = true, HPackSpecExamples.C61FirstResponseWithHuffman, None)
+        )
+        // Then the user can still consume the response
+        user.expectResponse()
+        user.responseIn.expectComplete()
+      }
+
+      "until in-flight streams deliver the request and responses are received" inAssertAllStagesStopped new TestSetup(
+        Setting(SettingIdentifier.SETTINGS_MAX_CONCURRENT_STREAMS, 3)
+      ) with NetProbes {
+        // Given several in-flight requests...
+        val request = HttpRequest(uri = "https://www.example.com/")
+        user.emitRequest(request)
+        user.emitRequest(request)
+        user.emitRequest(request)
+        // ... and backpressured requests...
+        user.emitRequest(request) // this emit succeeds but is buffered
+
+        network.expect[HeadersFrame]().streamId shouldBe (1)
+        network.expect[HeadersFrame]().streamId shouldBe (3)
+        network.expect[HeadersFrame]().streamId shouldBe (5)
+        network.expectNoBytes(100.millis)
+
+        // ... and some request already got their response (canary round-trip)
+        network.sendFrame(HeadersFrame(streamId = 1, endStream = true, endHeaders = true, HPackSpecExamples.C61FirstResponseWithHuffman, None))
+        val receivedResponse1: HttpResponse = user.expectResponse()
+
+        // ... and a completed user handler.
+        user.requestOut.sendComplete()
+
+        // When the first few response-frames arrive...
+        network.sendFrame(HeadersFrame(streamId = 3, endStream = true, endHeaders = true, HPackSpecExamples.C61FirstResponseWithHuffman, None))
+        // Then the backpressured in-flight streams proceed (sending pending requests)...
+        network.expect[HeadersFrame]().streamId shouldBe (7)
+        network.sendFrame(HeadersFrame(streamId = 5, endStream = true, endHeaders = true, HPackSpecExamples.C61FirstResponseWithHuffman, None))
+        network.sendFrame(HeadersFrame(streamId = 7, endStream = true, endHeaders = true, HPackSpecExamples.C61FirstResponseWithHuffman, None))
+        //... and it is possible to consume all responses (from stream 1 to stream 7)
+        val receivedResponse3: HttpResponse = user.expectResponse()
+        val receivedResponse5: HttpResponse = user.expectResponse()
+        val receivedResponse7: HttpResponse = user.expectResponse()
+      }
+
+      "until in-flight responses with chunked payload are fully received and consumed" ignore {
+        // Assert that not only the response but the full entity is available
+      }
```

Comment:
I've added this test skeleton as a response to https://github.com/akka/akka-http/pull/3734/files#r558348294.

We can merge this PR and implement the test on a new PR to get the feature in 10.2.3 or we can do things right and implement the test before merging this PR :-)

Reply:
What happens right now, it receives the headers/request but fails the body source?

Reply:
I think this should be covered by this PR as well.

Reply:
Isn't it already tested above? Response data is always streamed with HTTP/2 because headers and body arrive in distinct frames.

Reply:
Or is it about the distinction of reporting trailers with a chunked response?

Reply:
Looks like they're now tested above! (previously they weren't, this placeholder/TODO test can be removed now ;) )